src/platforms/common/blackpill-f4/blackpill-f4.c

```diff
@@ -173,24 +174,98 @@ void platform_target_clk_output_enable(bool enable)
 
 bool platform_spi_init(const spi_bus_e bus)
 {
-	(void)bus;
-	return false;
+	uint32_t controller = 0;
+	if (bus == SPI_BUS_INTERNAL) {
+		/* Set up onboard flash SPI GPIOs: PA5/6/7 as SPI1 in AF5, PA4 as nCS output push-pull */
+		gpio_mode_setup(OB_SPI_PORT, GPIO_MODE_AF, GPIO_PUPD_NONE, OB_SPI_SCLK | OB_SPI_MISO | OB_SPI_MOSI);
+		gpio_mode_setup(OB_SPI_PORT, GPIO_MODE_OUTPUT, GPIO_PUPD_NONE, OB_SPI_CS);
+		gpio_set_output_options(
+			OB_SPI_PORT, GPIO_OTYPE_PP, GPIO_OSPEED_50MHZ, OB_SPI_SCLK | OB_SPI_MISO | OB_SPI_MOSI | OB_SPI_CS);
+		gpio_set_af(OB_SPI_PORT, GPIO_AF5, OB_SPI_SCLK | OB_SPI_MISO | OB_SPI_MOSI);
+		/* Deselect the targeted peripheral chip */
+		gpio_set(OB_SPI_PORT, OB_SPI_CS);
+	} else if (bus == SPI_BUS_EXTERNAL) {
+		/* Set up external SPI GPIOs: PB13/14/15 as SPI2 in AF5, PB12 as nCS output push-pull */
+		gpio_mode_setup(EXT_SPI_PORT, GPIO_MODE_AF, GPIO_PUPD_NONE, EXT_SPI_SCLK | EXT_SPI_MISO | EXT_SPI_MOSI);
+		gpio_mode_setup(EXT_SPI_PORT, GPIO_MODE_OUTPUT, GPIO_PUPD_NONE, EXT_SPI_CS);
+		gpio_set_output_options(
+			EXT_SPI_PORT, GPIO_OTYPE_PP, GPIO_OSPEED_50MHZ, EXT_SPI_SCLK | EXT_SPI_MISO | EXT_SPI_MOSI | EXT_SPI_CS);
+		gpio_set_af(EXT_SPI_PORT, GPIO_AF5, EXT_SPI_SCLK | EXT_SPI_MISO | EXT_SPI_MOSI);
+		/* Deselect the targeted peripheral chip */
+		gpio_set(EXT_SPI_PORT, EXT_SPI_CS);
+	} else
+		return false;
+
+	if (bus == SPI_BUS_INTERNAL) {
+		rcc_periph_clock_enable(RCC_SPI1);
+		rcc_periph_reset_pulse(RST_SPI1);
+		controller = OB_SPI;
+	} else if (bus == SPI_BUS_EXTERNAL) {
+		rcc_periph_clock_enable(RCC_SPI2);
+		rcc_periph_reset_pulse(RST_SPI2);
+		controller = EXT_SPI;
+	}
+
+	/* Set up hardware SPI: master, PCLK/8, Mode 0, 8-bit MSB first */
+	spi_init_master(controller, SPI_CR1_BAUDRATE_FPCLK_DIV_8, SPI_CR1_CPOL_CLK_TO_0_WHEN_IDLE,
+		SPI_CR1_CPHA_CLK_TRANSITION_1, SPI_CR1_DFF_8BIT, SPI_CR1_MSBFIRST);
+	spi_enable(controller);
+	return true;
 }
 
 bool platform_spi_deinit(const spi_bus_e bus)
 {
-	(void)bus;
-	return false;
+	if (bus == SPI_BUS_INTERNAL) {
+		/* Notice: spi_clean_disable() hangs */
```

Comment:
Note that `spi_clean_disable()` is designed for when there could be stuff waiting in the data register, which is why it hangs, as we already ensure that not to be the case. Please remove this comment as we don't want to lead future readers to that function as it's explicitly the wrong tool here.

Reply:
This is remains of me trying to use libopencm3 APIs. Maybe I was mislead by their documentation, but it actually hangs the probe with no timeouts, I didn't expect it to. Will remove the commit

Reply:
They don't actually mention, to be fair, that it will hang if there's no data in the DR to read, but.. it'll be hanging on the RXNE check:

https://github.com/blackmagic-debug/libopencm3/blob/0c6c7c2895c1c2b81c157242a754f99964de828d/lib/stm32/common/spi_common_all.c#L104-L120